next.mdx.shiki.mjs

```diff
@@ -19,8 +19,150 @@ const memoizedShikiji = await getHighlighterCore({
 // to attribute the current language of the <pre> element
 const languagePrefix = 'language-';
 
+/**
+ * Retrieve the value for the given meta key.
+ *
+ * @example - Returns "CommonJS"
+ * getMetaParameter('displayName="CommonJS"', 'displayName');
+ *
+ * @param {any} meta - The meta parameter.
+ * @param {string} key - The key to retrieve the value.
+ *
+ * @return {string | undefined} - The value related to the given key.
+ */
+function getMetaParameter(meta, key) {
+  if (typeof meta !== 'string') {
+    return;
+  }
+
+  const matches = meta.match(new RegExp(`${key}="(?<parameter>[^"]*)"`));
+  const parameter = matches?.groups.parameter;
+
+  return parameter !== undefined && parameter.length > 0
+    ? parameter
+    : undefined;
+}
+
+/**
+ * @typedef {import('unist').Node} Node
+ * @property {string} tagName
+ * @property {Node[]} children
+ */
+
+/**
+ * Checks if the given node is a valid code element.
+ *
+ * @param {Node} node - The node to be verified.
+ *
+ * @return {boolean} - True when it is a valid code element, false otherwise.
+ */
+function isCodeBlock(node) {
+  return node?.tagName === 'pre' && node?.children[0].tagName === 'code';
+}
+
+/**
+ * Retrieves a list indicating the starting, and ending indexes of sequential
+ * code elements.
+ *
+ * @param {Node} tree - The current MDX resolved content.
+ *
+ * @return {{start: number, end: number}[]} - The list containing every range of
+ * sequential code elements.
+ */
+function getCodeTabsRange(tree) {
+  const rangeMap = {};
+  let start = null;
+
+  visit(tree, 'element', (node, index, parent) => {
+    // Adding 2 since there is one text node between every element
+    const next = index + 2;
+
+    if (
+      isCodeBlock(node) &&
+      'children' in parent &&
+      parent.children !== undefined &&
+      isCodeBlock(parent.children[next])
+    ) {
+      start ??= index;
+      rangeMap[start] = next;
+
+      // Prevent visiting the code block children
+      return SKIP;
+    }
+
+    // End of sequential code elements, reset the start for the next range
+    start = null;
+  });
+
+  return Object.entries(rangeMap).map(([start, end]) => ({
+    start: Number(start),
+    end: Number(end),
+  }));
+}
+
 export default function rehypeShikiji() {
   return async function (tree) {
+    // Retrieve all sequential code boxes to transform
+    const ranges = getCodeTabsRange(tree);
```

Comment:
You're aware this change will have an immediate effect on the legacy website right? I wonder what the penalty in performance will be and what changes this might imply.

Reply:
Anyhow, what worries me the most here, is that with this change, we will "visit" the whole tree at least N times.

I understand why this is being done, but it feels **fragile** My worry here is that we might be overcomplicating everything here.

I wonder if instead of having all of this complexity, we could visit the root tree and simply go down:

-> root tree
  -> detects pre element
     -> goes down that pre element
       -> transforms into shiki
       -> add the resulting element to a temporary array
     -> seeks if there is another pre element
       -> if yes, same logic as below
  -> once goes all close codeboxes
     -> replaces the elements with the contents of the temporary array (the whole range)
  -> keeps going down the tree, identifying pre elements

With this IMO, we only need one `visit` call. Remember that `visit` goes by every element of the tree by itself, we don't need to call `visit` on deeper items of the tree... I think.

I just believe we can optimise this further, as I'm truly fearful of issues.